### PR TITLE
Non ML Baseline and Overall Analysis

### PR DESCRIPTION
**Pull request by rhadi-moz**
# Overview 
This merge request contains the following changes: 
 
## Add comparison with "normalized" url 
- Based on initial review of retrieval, there are similar URLs (e.g. google mail variants, different 'edit' states of the same google doc) that were negatively counting toward the retrieval. Now, we have the ability to "normalize" the URL to allow comparison to the more root forms, while still ensuring they represent distinct pages.  
- The retrieval itself is the same, but in the construction of the ground truth, we add the normalized URL 
- For evaluation, we add another set of the same evaluation metrics, this time comparing against `norm_relevant_docs` (the expanded list) 
- Because we are only doing k=2, this will make recall look worse (because there are many variants of the 'same URL' not retrieved. It is expected to improve precision @ 2 and mean average precision 

## Understand lexical overlap of search query + retrieved 
Similar to the idea of exact match, added some metrics to help us understand the overlap of the search query with the 'combined_text'. Part of the goal with semantic search is to augment search when there is no/minimal keyword overlap (strategy, approach) <-- low lexical overlap, similar meaning, or when additional words change the meaning or context (e.g. work from home, not working from home) <-- high lexical overlap, opposite semantic meaning 

## Add comparison to non ML baseline 
- To contextualize performance of embedding models, added a script that does tf-idf, BM25, and exact match retrieval 
- Exact match was a bit tricky (is it exact match at the entire query? sub terms of query?). To keep it simple, I added both versions, the latter of which I also sort by frecency descending (the goal is to approximate current suggest functionality) 

## Add Overall Model Comparison 
- Add a script that takes all the evaluation results and creates plots of them for each metric, to make analysis easier

### CODE REVIEW

File: evaluation_pipeline/evaluation.py [hunk 1]
```diff
@@ -163,6 +161,8 @@ def run_traditional_eval(query_id, query, relevant_docs, retrieved_docs, retriev
     ndcg = calc_ndcg(relevant_docs, retrieved_docs,score_type='rank',retrieved_distances=retrieved_distances, k=k)
     reciprocal_rank = calc_reciprocal_rank(relevant_docs, retrieved_docs)
     average_precision = calc_average_precision(relevant_docs, retrieved_docs, k=k)
+    jaccard_indices = [calc_jaccard_index(query, text) for text in retrieved_texts]
```
Comment (chidamnat) [inline evaluation_pipeline/evaluation.py:164]: nit - text may have stop words and that could potentially affect the jaccard index and overlap_coefficient

File: non_ml_comparison.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+def normalize_url(url):
+    parsed = urlparse(url)
+    # Remove fragments and queries, keep scheme + netloc + path
+    return urlunparse((parsed.scheme, parsed.netloc, parsed.path, '', '', ''))
+
+
+
+def process_data(history_file_path, query_file_path):
+    # Define Corpus and Queries
+    history = process_history(row_limit=10001, history_file_path=history_file_path)
```
Comment (chidamnat) [inline non_ml_comparison.py:26]: could we pls create a constant for ROW_LIMIT = 10001 so code change will be easier

File: non_ml_comparison.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        list[dict]: A list of dictionaries containing the retrieval results for each query.
+    """
+    results = {}
+
+    for i, row in unique_search_query_df.iterrows():
+        query_id = row['query_id']
+        query = row['search_query'].lower()
+       # query_terms = query.split()  # Split query into words
+        #pattern = r"\b(" + "|".join(query_terms) + r")\b"
+        #matches = history[history['combined_text'].str.contains(pattern, case=False, na=False, regex=True)]
```
Comment (chidamnat) [inline non_ml_comparison.py:147]: commented lines could be dropped if not used

File: non_ml_comparison.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+    for i, row in unique_search_query_df.iterrows():
+        query_id = row['query_id']
+        query = row['search_query'].lower()
+        query_terms = query.split()  # Split query into words
+        pattern = r"\b(" + "|".join(query_terms) + r")\b"
+        matches = history[history['combined_text'].str.contains(pattern, case=False, na=False,regex=True)].sort_values(by='frecency', ascending=False)
+
+        # Filter documents in the history DataFrame where `combined_text` contains the exact query terms
+#        matches = history[history['combined_text'].str.contains(rf'\b{query}\b', case=False, na=False)]
```
Comment (chidamnat) [inline non_ml_comparison.py:194]: same as above

File: non_ml_comparison.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.metrics.pairwise import cosine_similarity
+from rank_bm25 import BM25Okapi
+from nltk.tokenize import word_tokenize
```
Comment (chidamnat) [inline non_ml_comparison.py:4]: I think we need these new dependencies in requirements.txt. could you pls include evaluation_pipeline/requirements.txt

File: non_ml_comparison.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+            for _, doc in matches.iterrows()
+        ]
+
+        results[query_id] =  retrievals
+
+    return results
+
+
+
+def exact_match_terms_retrieval_frecency(history, unique_search_query_df, top_k):
```
Comment (chidamnat) [inline non_ml_comparison.py:172]: looks like lot of similarity with `exact_match_retrieval` function. could we pls combine them or repurpose

**Comment by chidamnat** [comment]
could you pls rebase with main. the data path was fixed and merged in main, that could be reflect by this branch. makes it easy for test

**Comment by rhadi-moz** [comment]
Closing this PR b/c I did enough refactoring to warrant a new one to hopefully make changes more clear
